File: 3144. Minimum Substring Partition of Equal Character Frequency.java
class Solution {
    public int minimumSubstringsInPartition(String s) {
        Integer[] cache = new Integer[s.length()];
        return helper(s, 0, cache);
    }
    private int helper(String s, int index, Integer[] cache) {
        if (index == s.length()) return 0;
        if (cache[index] != null) return cache[index];
        int[] freq = new int[26];
        int answer = s.length();
        for (int i = index; i < s.length(); i++) {
            int c = s.charAt(i) - 'a';
            freq[c] += 1;
            if (check(freq)) {
                answer = Math.min(answer, 1 + helper(s, i+1, cache));
            }
        }
        return cache[index] = answer;
    }
    private boolean check(int[] freq) {
        int min = 100000, max = -1;
        for (int i : freq) {
            if (i > 0) {
                min = Math.min(min, i);
                max = Math.max(max, i);
            }
        }
        return min == max;
    }
}

File: 3144. Minimum Substring Partition of Equal Character Frequency.go
func minimumSubstringsInPartition(s string) int {
	cache := make([]int, len(s))
	for i := range cache {
		cache[i] = -1
	}
	return helper(s, 0, cache)
}
func helper(s string, index int, cache []int) int {
	if index == len(s) {
		return 0
	}
	if cache[index] != -1 {
		return cache[index]
	}
	freq := make([]int, 26)
	answer := len(s)
	for i := index; i < len(s); i++ {
		freq[s[i]-'a']++
		if check(freq) {
			answer = min(answer, 1+helper(s, i+1, cache))
		}
	}
	cache[index] = answer
	return answer
}
func check(freq []int) bool {
	min, max := 100000, -1
	for _, f := range freq {
		if f == 0 {
			continue
		}
		if f < min {
			min = f
		}
		if f > max {
			max = f
		}
	}
	return min == max
}